File: economics.py
#!/usr/bin/env python3
"""Unit economics, derived from the finished schedule.

Reference implementation. src/economics.js mirrors this, the same way
validate.js mirrors validate.py -- if the two disagree that is a real defect.

This reads the schedule and costs it. It never feeds back into the engine, so
the eight acceptance metrics and the ten checks cannot move because of
anything in here.

    python3 make_legs.py && python3 engine.py && python3 economics.py

WHAT MAKES STAGE LENGTH BITE. Every element is stored as a per-cycle plus a
per-hour rate (see tools/f41_rates.py for how the split is derived, and why
Form 41 cannot measure it). A 300 nm E145 leg pays a full cycle for very few
hours; a transcon amortises the cycle over five. That is the entire reason the
model can tell you a short leg is uneconomic.

MARGINAL VERSUS FULLY ALLOCATED. Ownership is a fixed cost of a roster that
already exists. Charging a route for an aircraft it did not cause is the
classic way to reject a route that would have made money, so `direct` excludes
ownership and `allocated` includes it. For "should this route exist", the
honest test is direct cost against revenue, plus an aircraft only when a
rebuild says the fleet requirement actually moved.
"""
import json, math, pathlib
from collections import defaultdict

ROOT = pathlib.Path(__file__).resolve().parent
load = lambda p: json.loads((ROOT / p).read_text())

ECON = load("src/data/economics.json")
FLEETDATA = load("src/data/fleet.json")
FLEET = {f["t"]: f for f in FLEETDATA["types"]}
AP = load("src/data/airports.json")
LINES = load("lines.json")["lines"]

RATES = ECON["types"]
FUEL = ECON["fuel"]["pricePerGal"]
STATION = ECON["station"]
OPS = ECON.get("ops", {})
SPARE = OPS.get("spareRatio", 0.08)
OVERHEAD = OPS.get("overheadPct", 0.0)
HANDLING = OPS.get("handlingPerDep", 0.0)
REF_SEATS = OPS.get("landingRefSeats", 175)


def rate(typ):
    """Form 41 rates for one of this airline's fleet types.

    `f41` on the fleet type says which real aircraft it is costed as; `f41Adj`
    multiplies individual elements where the airline's variant differs. The
    A320E is an A320 with international handling, not a separate cost basis.
    """
    f = FLEET[typ]
    r = RATES.get(f.get("f41", typ))
    if not r:
        raise SystemExit(f"no Form 41 rates for fleet type {typ} "
                         f"(f41={f.get('f41', typ)!r})")
    adj = f.get("f41Adj", {})
    out = {}
    for k in ("gal", "crew", "maint"):
        m = adj.get(k, 1.0)
        out[k] = {"cyc": r[k]["cyc"] * m, "hr": r[k]["hr"] * m}
    out["ownDay"] = (r["ownDay"] or 0) * adj.get("ownDay", 1.0)
    return out


def cost_of_flight(o, d, typ, blk_hr):
    """Direct cost of one departure. Ownership is deliberately not in here."""
    r = rate(typ)
    seats = FLEET[typ]["F"] + FLEET[typ]["PE"] + FLEET[typ]["Y"]
    gal = r["gal"]["cyc"] + r["gal"]["hr"] * blk_hr
    c = {
        "fuel": gal * FUEL,
        "crew": r["crew"]["cyc"] + r["crew"]["hr"] * blk_hr,
        "maint": r["maint"]["cyc"] + r["maint"]["hr"] * blk_hr,
        # Landing fees scale with weight; seats are the only weight proxy the
        # model carries, and the peer figure is at the peer average gauge.
        "landing": STATION["landingPerDep"] * seats / REF_SEATS,
        "handling": HANDLING,
    }
    c["gal"] = gal
    c["direct"] = sum(c[k] for k in
                      ("fuel", "crew", "maint", "landing", "handling"))
    return c


def build():
    """Cost every flight in the schedule, then roll up."""
    flights = []
    block_by_type = defaultdict(float)
    tails = defaultdict(int)
    for l in LINES:
        tails[l["type"]] += 1
        for o, d, dep, arr, b in l["flights"]:
            blk = b / 60.0
            c = cost_of_flight(o, d, l["type"], blk)
            block_by_type[l["type"]] += blk
            flights.append(dict(o=o, d=d, t=l["type"], base=l["base"],
                                blk=blk, nm=dist(o, d), cost=c))

    # Ownership is a daily charge on the whole roster, spares included, spread
    # across that type's flying. High utilisation is what makes it cheap per
    # departure, which is why it cannot be a per-departure constant.
    own_rate = {}
    for t, n in tails.items():
        total = math.ceil(n * (1 + SPARE))
        bh = block_by_type[t]
        own_rate[t] = (rate(t)["ownDay"] * total / bh) if bh else 0.0
    for f in flights:
        f["cost"]["own"] = own_rate[f["t"]] * f["blk"]
        f["cost"]["allocated"] = (f["cost"]["direct"] + f["cost"]["own"]) \
            * (1 + OVERHEAD)
    return flights, tails, own_rate


DIST = {}


def dist(a, b):
    k = (a, b) if a < b else (b, a)
    if k not in DIST:
        la1, lo1 = math.radians(AP[a][2]), math.radians(AP[a][3])
        la2, lo2 = math.radians(AP[b][2]), math.radians(AP[b][3])
        DIST[k] = 3440.065 * 2 * math.asin(math.sqrt(
            math.sin((la2 - la1) / 2) ** 2 + math.cos(la1) * math.cos(la2)
            * math.sin((lo2 - lo1) / 2) ** 2))
    return DIST[k]


def report(flights, tails, own_rate):
    SM = 1.15078
    seats = lambda t: FLEET[t]["F"] + FLEET[t]["PE"] + FLEET[t]["Y"]
    asm = sum(seats(f["t"]) * f["nm"] * SM for f in flights)
    tot = defaultdict(float)
    for f in flights:
        for k, v in f["cost"].items():
            tot[k] += v
    print(f"{len(flights)} flights, {sum(tails.values())} tails, "
          f"{asm/1e6:.1f}m ASM/day\n")
    print(f"{'element':12}{'$/day':>14}{'cents/ASM':>11}{'share':>8}")
    for k in ("fuel", "crew", "maint", "landing", "handling", "own"):
        print(f"{k:12}{tot[k]:14,.0f}{tot[k]/asm*100:11.2f}"
              f"{tot[k]/tot['allocated']*100:7.0f}%")
    print(f"{'-'*45}")
    print(f"{'direct':12}{tot['direct']:14,.0f}{tot['direct']/asm*100:11.2f}")
    print(f"{'allocated':12}{tot['allocated']:14,.0f}"
          f"{tot['allocated']/asm*100:11.2f}")

    print(f"\n{'type':7}{'deps':>6}{'stage':>7}{'blk/dep':>9}{'$/dep':>9}"
          f"{'CASM':>7}{'seat-mi':>9}")
    by = defaultdict(lambda: defaultdict(float))
    for f in flights:
        a = by[f["t"]]
        a["deps"] += 1
        a["nm"] += f["nm"]
        a["blk"] += f["blk"]
        a["cost"] += f["cost"]["allocated"]
        a["asm"] += seats(f["t"]) * f["nm"] * SM
    for t in FLEET:
        a = by.get(t)
        if not a:
            continue
        print(f"{t:7}{a['deps']:6.0f}{a['nm']/a['deps']:7.0f}"
              f"{a['blk']/a['deps']:9.2f}{a['cost']/a['deps']:9,.0f}"
              f"{a['cost']/a['asm']*100:7.2f}{a['asm']/a['deps']:9,.0f}")

    # CASM against stage length -- the curve the whole model exists to show
    print(f"\n{'stage band':14}{'deps':>7}{'CASM':>8}{'$/dep':>9}  types")
    bands = [(0, 300), (300, 500), (500, 750), (750, 1000), (1000, 1500),
             (1500, 9999)]
    for lo, hi in bands:
        fs = [f for f in flights if lo <= f["nm"] < hi]
        if not fs:
            continue
        c = sum(f["cost"]["allocated"] for f in fs)
        a = sum(seats(f["t"]) * f["nm"] * SM for f in fs)
        ts = sorted({f["t"] for f in fs})
        label = f"{lo}-{hi} nm" if hi < 9999 else f"{lo}+ nm"
        print(f"{label:14}{len(fs):7}{c/a*100:8.2f}{c/len(fs):9,.0f}  "
              f"{','.join(ts)}")


def totals(flights, tails, own_rate):
    SM = 1.15078
    seats = lambda t: FLEET[t]["F"] + FLEET[t]["PE"] + FLEET[t]["Y"]
    tot = defaultdict(float)
    for f in flights:
        for k, v in f["cost"].items():
            tot[k] += v
        tot["asm"] += seats(f["t"]) * f["nm"] * SM
    tot["deps"] = len(flights)
    return dict(tot)


if __name__ == "__main__":
    import sys
    built = build()
    if "--json" in sys.argv:
        t = totals(*built)
        (ROOT / "econ_py.json").write_text(json.dumps(
            {"totals": t, "casm": t["allocated"] / t["asm"] * 100}, indent=1))
        print("wrote econ_py.json")
    else:
        report(*built)

File: src/economics.js
/* ---------- economics ----------
   Unit costs derived from the finished schedule, the same way validate.js
   re-derives the ten checks: this reads M.flights and costs them, and never
   feeds anything back into the engine. The eight acceptance metrics and the
   ten checks cannot move because of anything in this file.

   Mirrors economics.py. Any disagreement between the two is a real defect.

   Rates come from src/data/economics.json, extracted from published DOT
   Form 41 by tools/f41_rates.py. Every element is a per-cycle plus a per-hour
   rate, which is what makes stage length change unit cost -- a 300 nm leg
   pays a whole cycle for 1.3 block hours, a transcon spreads it over five.
   Form 41 cannot measure that split (the cross-carrier regression returns
   negative cycle fuel), so the level is measured and only the shape is
   assumed, through cycleEquivHours in the JSON. */

const ECON = (()=>{ try { return JSON.parse(document.getElementById("econ").textContent); }
                    catch(e){ console.error("economics data unreadable", e);
                              return {types:{}, fuel:{pricePerGal:0},
                                      station:{landingPerDep:0}, ops:{}}; } })();
const ECON_OPS = ECON.ops || {};
const OVERHEAD_PCT   = ECON_OPS.overheadPct     != null ? ECON_OPS.overheadPct : 0;
const HANDLING_DEP   = ECON_OPS.handlingPerDep  != null ? ECON_OPS.handlingPerDep : 0;
const LAND_REF_SEATS = ECON_OPS.landingRefSeats != null ? ECON_OPS.landingRefSeats : 175;

/* Form 41 rates for one of this airline's fleet types. `f41` on the fleet type
   says which real aircraft it is costed as, `f41Adj` scales individual elements
   where the variant differs — the A320E is an A320 doing international work,
   not a separate cost basis. Returns null if the type has no rates, so the tab
   can say which types are uncosted instead of quietly reporting zero. */
function econRate(t){
  const spec = SPEC[t] || {}, base = FLEET_BASE.find(f=>f.t===t) || {};
  const key = base.f41 || spec.f41 || t, r = ECON.types[key];
  if(!r) return null;
  const adj = base.f41Adj || spec.f41Adj || {}, out = {f41:key};
  for(const k of ["gal","crew","maint"]){
    const m = adj[k] != null ? adj[k] : 1;
    out[k] = {cyc: r[k].cyc*m, hr: r[k].hr*m};
  }
  out.ownDay = (r.ownDay||0) * (adj.ownDay != null ? adj.ownDay : 1);
  out.flags = r.flags || [];
  return out;
}

/* Direct cost of one departure. Ownership is deliberately excluded: it is a
   fixed cost of a roster that already exists, and charging a candidate route
   for an aircraft it did not cause is how you reject a route that would have
   made money. See econMarginal() for the honest test. */
function econFlightCost(t, blkHr){
  const r = econRate(t);
  if(!r) return null;
  const seats = (SPEC[t] && SPEC[t].seats) || 0;
  const gal = r.gal.cyc + r.gal.hr*blkHr;
  const c = {
    gal,
    fuel:  gal * ECON.fuel.pricePerGal,
    crew:  r.crew.cyc  + r.crew.hr*blkHr,
    maint: r.maint.cyc + r.maint.hr*blkHr,
    // Landing fees scale with weight; seats are the only weight proxy the
    // model carries, and the peer figure is at the peer average gauge.
    landing: ECON.station.landingPerDep * seats / LAND_REF_SEATS,
    handling: HANDLING_DEP
  };
  c.direct = c.fuel + c.crew + c.maint + c.landing + c.handling;
  return c;
}

/* Cost the whole schedule. Returns per-flight costs plus rollups by type,
   by route and by stage-length band. */
function econModel(M){
  const flights = [], byType = new Map(), blockByType = new Map(),
        tails = new Map(), uncosted = new Set();
  for(const r of M.rots) tails.set(r.t, (tails.get(r.t)||0)+1);
  for(const f of M.flights){
    const c = econFlightCost(f.t, f.blk);
    if(!c){ uncosted.add(f.t); continue; }
    blockByType.set(f.t, (blockByType.get(f.t)||0) + f.blk);
    flights.push({f, cost:c});
  }
  /* Ownership is a daily charge on the whole roster, spares included, spread
     across that type's flying. Utilisation is what makes it cheap per
     departure, so it cannot be a per-departure constant. */
  const spare = state.spare === undefined ? 0.08 : state.spare, ownHr = new Map();
  for(const [t,n] of tails){
    const r = econRate(t), bh = blockByType.get(t)||0;
    ownHr.set(t, (r && bh) ? r.ownDay*Math.ceil(n*(1+spare))/bh : 0);
  }
  const tot = {fuel:0, crew:0, maint:0, landing:0, handling:0, own:0,
               direct:0, allocated:0, gal:0, asm:0, deps:0, blk:0};
  for(const x of flights){
    const {f, cost} = x;
    cost.own = (ownHr.get(f.t)||0) * f.blk;
    cost.allocated = (cost.direct + cost.own) * (1 + OVERHEAD_PCT);
    const asm = ((SPEC[f.t] && SPEC[f.t].seats)||0) * f.nm * SM;
    for(const k of ["fuel","crew","maint","landing","handling","own","direct",
                    "allocated","gal"]) tot[k] += cost[k];
    tot.asm += asm; tot.deps++; tot.blk += f.blk;
    let a = byType.get(f.t);
    if(!a){ a = {t:f.t, deps:0, nm:0, blk:0, asm:0, direct:0, allocated:0};
            byType.set(f.t, a); }
    a.deps++; a.nm += f.nm; a.blk += f.blk; a.asm += asm;
    a.direct += cost.direct; a.allocated += cost.allocated;
  }
  // by market, both directions together, since a turn is one aircraft
  const byRoute = new Map();
  for(const {f, cost} of flights){
    const k = pairKey(f.o, f.d);
    let a = byRoute.get(k);
    if(!a){ a = {key:k, o:f.o, d:f.d, nm:f.nm, deps:0, seats:0, asm:0,
                 direct:0, allocated:0, types:new Set()};
            byRoute.set(k, a); }
    a.deps++; a.types.add(f.t);
    a.seats += (SPEC[f.t] && SPEC[f.t].seats)||0;
    a.asm += ((SPEC[f.t] && SPEC[f.t].seats)||0) * f.nm * SM;
    a.direct += cost.direct; a.allocated += cost.allocated;
  }
  const bands = [[0,300],[300,500],[500,750],[750,1000],[1000,1500],[1500,1e9]];
  const byStage = bands.map(([lo,hi])=>{
    const fs = flights.filter(x=>x.f.nm>=lo && x.f.nm<hi);
    const asm = fs.reduce((s,x)=>s+((SPEC[x.f.t]&&SPEC[x.f.t].seats)||0)*x.f.nm*SM, 0);
    const cost = fs.reduce((s,x)=>s+x.cost.allocated, 0);
    return {lo, hi, deps:fs.length, asm, cost,
            casm: asm ? cost/asm*100 : 0, perDep: fs.length ? cost/fs.length : 0};
  }).filter(b=>b.deps);
  return {flights, totals:tot, byType:[...byType.values()],
          byRoute:[...byRoute.values()], byStage,
          casm: tot.asm ? tot.allocated/tot.asm*100 : 0,
          uncosted:[...uncosted], ownHr};
}

/* The honest cost of adding a route: its own direct cost, plus an aircraft
   only if a full rebuild says the fleet requirement actually moved. The app is
   unusually able to answer that, because trialBuild() already rebuilds the
   whole schedule for every suggestion candidate. `deltaTails` comes from that
   comparison; 0 means the route fitted into existing slack and owes nothing. */
function econMarginal(before, after){
  const b = econModel(before), a = econModel(after);
  // Tails by type, so the aircraft charge is the type that actually grew
  // rather than an average of the fleet.
  const count = M => { const m = new Map();
    for(const r of M.rots) m.set(r.t, (m.get(r.t)||0)+1); return m; };
  const cb = count(before), ca = count(after);
  let deltaTails = 0, ownDay = 0;
  for(const t of new Set([...cb.keys(), ...ca.keys()])){
    const n = (ca.get(t)||0) - (cb.get(t)||0);
    if(!n) continue;
    deltaTails += n;
    const r = econRate(t); if(r) ownDay += n * r.ownDay;
  }
  const deltaDirect = a.totals.direct - b.totals.direct;
  return {
    deltaDirect, deltaTails, ownDay,
    // What the route actually owes: its own operating cost, plus an aircraft
    // only when the rebuild says one appeared.
    marginal: deltaDirect + ownDay,
    deltaAllocated: a.totals.allocated - b.totals.allocated,
    deltaDeps: a.totals.deps - b.totals.deps,
    deltaAsm: a.totals.asm - b.totals.asm
  };
}
